### PR TITLE
Add test for metrics and tracing endpoints

### LINKED ISSUE #995: Add a metrics and tracing listener test to the zebrad acceptance tests
**Is your feature request related to a problem? Please describe.**

`zebrad`'s acceptance tests don't open metrics and tracing listeners by default.

We should test those code paths.

**Describe the solution you'd like**

Add a separate test to `zebrad/tests/acceptance.rs` which runs `zebrad start` with metrics and tracing listeners on `127.0.0.1:0`. (A zero port means "pick any available port.)

Use the default test config, and modify it to add listeners, like the existing `misconfigured_ephemeral_mode` test:
https://github.com/ZcashFoundation/zebra/blob/main/zebrad/tests/acceptance.rs#L329

The test should check that the command does not panic or exit. Because we expect the command to continue running, the test should kill the command, and check that it was killed, like:
https://github.com/ZcashFoundation/zebra/blob/main/zebrad/tests/acceptance.rs#L342 

**Describe alternatives you've considered**

A. Also add an IPv6 `[::1]:0` test.

But we can't be sure that all our users will have IPv6, so:
* the tests would need to check for "no IPv6" or "keeps running", and
* the "no IPv6" message could be different on different OSes.

Since zebrad doesn't handle raw IP addresses, it is unlikely that we'll write code which works for IPv4, but not IPv6. So this test is a low priority.

B. Do not add any extra tests.

These code paths would not be tested.

**Additional context**

We tried a more complex version of this test in #951, along with other related changes.

### PR DESCRIPTION
**Pull request by oxarbitrage**
Resolves https://github.com/ZcashFoundation/zebra/issues/995

### CODE REVIEW

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: before we kill the process, could we try connecting to each endpoint and do some basic interactions with them like setting the filter and then looking it up.

**Review by yaahc — CHANGES_REQUESTED** [review]
looks good, just want to add a couple more checks

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: What do you think it will be the best crate to be used for connecting to the endpoints ? `std::net::TcpStream` was the natural choice for me however not totally sure, there are other options.

Comment (yaahc) [reply zebrad/tests/acceptance.rs]: I think we already have a dependency on `hyper` so that would probably be best, I think doing it via `TcpStream` directly might force you to deal with http headers, not sure though.

Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: `hyper` looks good to me. Thanks.

Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Applied partially at https://github.com/ZcashFoundation/zebra/pull/1000/commits/21387745f5729d63cffb1b2b23eafa503e30d11b

Set the filter is blocked by https://github.com/ZcashFoundation/zebra/issues/1001

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]:
```suggestion
    // Run the program for a second before testing the endpoints
```

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: do we need to create a second `Client`?

**Review by yaahc — CHANGES_REQUESTED** [review]
the hyper additions look good, I think go ahead and apply the changes that test the endpoint even tho it will make the test fail. We won't be able to merge it until we fix the endpoint but having the test will make it much easier to fix the endpoint.

File: zebrad/tests/acceptance.rs
Comment (yaahc) [reply zebrad/tests/acceptance.rs]: makes sense, replied in my review comment https://github.com/ZcashFoundation/zebra/pull/1000#pullrequestreview-482941151

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Oops, thanks. Fixed this as part of https://github.com/ZcashFoundation/zebra/pull/1000/commits/939349cb9413391f2564a787246f8a1e0c96e3dc

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: No need, changed to 1 client as part of https://github.com/ZcashFoundation/zebra/pull/1000/commits/939349cb9413391f2564a787246f8a1e0c96e3dc

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [inline zebrad/tests/acceptance.rs:563]: Test will be locked here(#1001). This is the same as doing `curl -X POST localhost:3000/filter -d "zebrad=trace` as the docs state(https://github.com/ZcashFoundation/zebra/blob/main/zebrad/src/components/tracing/endpoint.rs#L91). When we do that the endpoint is not returning.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Can you split this test into separate metrics and tracing tests?

We try to test one feature per acceptance test, so it's easier to tell which feature is failing.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Consider choosing ports from 50000-60000, because applications usually don't bind to those port ranges:
https://en.m.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers

Also, let's add a note about how to fix port conflicts.
There is already a comment about port conflicts in another test:
https://github.com/ZcashFoundation/zebra/blob/main/zebrad/tests/acceptance.rs#L447

Let's move that note to the module comment at the top of the file. Then we can refer to it from all 3 tests.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: If we're going to depend on this default, maybe we should explicitly set it?

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Trace might produce lots of output, let's just use debug?
```suggestion
        .body(Body::from("zebrad=debug"))
```

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs:579]: Good idea!

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks great, thanks for doing this work!

I have a suggestion for moving forward:
1. Split the metrics and tracing tests
2. Disable the tracing test by default using `#[ignore]`
3. Merge this PR
4. When #1001 merges, enable the tracing test by default

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Done at https://github.com/ZcashFoundation/zebra/pull/1000/commits/e301ab938643fe8369752e9228fda5094699efbb

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: https://github.com/ZcashFoundation/zebra/pull/1000/commits/e301ab938643fe8369752e9228fda5094699efbb

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Another option is to test the root address instead of the tracing level. This was done as part of https://github.com/ZcashFoundation/zebra/pull/1000/commits/e301ab938643fe8369752e9228fda5094699efbb

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Done as part of https://github.com/ZcashFoundation/zebra/pull/1000/commits/e301ab938643fe8369752e9228fda5094699efbb

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Let's document the conflict where the error occurs:
```suggestion
    // [Note on port conflict](#Note on port conflict)
```

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Let's document the conflict where the error occurs:
```suggestion
    // [Note on port conflict](#Note on port conflict)
```

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: This isn't your bug, but we don't want to replace this error - we just want to wrap it with an extra hint.

We'll need to use eyre::WrapErr to use this function.

```suggestion
    output.assert_was_killed().wrap_err("Possible port or cache conflict. Are there other acceptance test, zebrad, or zcashd processes running?")?;
```

Comment (yaahc) [reply zebrad/tests/acceptance.rs]: expect won't replace the error, it adds a top level error and prints the value of the error being unwrapped, so this is fine as is imo, that said, wrap_err is probably better so it doesn't look like two errors being throw at once (once I also merge my panic_message customization stuff to supress the panic error message from tests that return Result::Err.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Let's add a useful hint to the error message.

We'll need to use eyre::WrapErr to use this function.

```suggestion
    output.assert_was_killed().wrap_err("Possible port conflict. Are there other acceptance tests running?")?;
```

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Let's add a useful hint to the error message.

We'll need to use eyre::WrapErr to use this function.

```suggestion
    output.assert_was_killed().wrap_err("Possible port conflict. Are there other acceptance tests running?")?;
```

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks great!

Just a few minor tweaks, then we're done.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [reply zebrad/tests/acceptance.rs:563]: I added "enable the tracing_endpoint acceptance test by default" to the TODO list in #1001

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: https://github.com/ZcashFoundation/zebra/pull/1000/commits/4e6bca466a78290c4a4d9e2906521cdcaf66ca54

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: https://github.com/ZcashFoundation/zebra/pull/1000/commits/4e6bca466a78290c4a4d9e2906521cdcaf66ca54

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: https://github.com/ZcashFoundation/zebra/pull/1000/commits/4e6bca466a78290c4a4d9e2906521cdcaf66ca54

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: https://github.com/ZcashFoundation/zebra/pull/1000/commits/4e6bca466a78290c4a4d9e2906521cdcaf66ca54

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: https://github.com/ZcashFoundation/zebra/pull/1000/commits/4e6bca466a78290c4a4d9e2906521cdcaf66ca54

**Review by teor2345 — APPROVED** [review]
Looks good to me, @yaahc are you ok with these changes?

We disabled the failing test by default.

**Review by yaahc — APPROVED** [review]
looks good to me

**Pull request merged**
